### PR TITLE
Fix BaseStream buf io concurrent access.

### PR DESCRIPTION
**Pull request by Frozen**
## Issue

Bufio Reader/Writer is not thread safe, different goroutines can work with BaseStream at the same time.

## Operational Checklist

1. **Does this PR introduce backward-incompatible changes to the on-disk data structure and/or the over-the-wire protocol?**. (If no, skip to question 8.)

    **NO**

8. **Does this PR introduce backward-incompatible changes *NOT* related to on-disk data structure and/or over-the-wire protocol?** (If no, continue to question 11.)

    **NO**

11. **Does this PR introduce significant changes to the operational requirements of the node software, such as >20% increase in CPU, memory, and/or disk usage?**

    **NO**

### CODE REVIEW

**Review by LeoHChen — APPROVED** [review]

**Comment by LeoHChen** [comment]
do you have a testcase that can trigger the concurrency issue?

File: p2p/stream/types/stream.go [hunk 1]
```diff
@@ -89,18 +90,19 @@ func (st *BaseStream) WriteBytes(b []byte) (err error) {
...
-	}
-	bytesWriteCounter.Add(sizeBytes)
-	if _, err = st.rw.Write(b); err != nil {
-		return
+	size := sizeBytes + len(b)
+	message := make([]byte, size)
+	copy(message, intToBytes(len(b)))
+	copy(message[sizeBytes:], b)
+
+	if _, err = st.raw.Write(message); err != nil {
```
Comment (JackyWYX) [inline p2p/stream/types/stream.go:98]: Shall we also add the lock to Write method?

**Comment by JackyWYX** [comment]
Thanks for pointing this out! Please also add a lock to write method to ensure the atomacity in

```
st.raw.Write
```

File: p2p/stream/types/stream.go (hunk 1 above)
Comment (Frozen) [reply p2p/stream/types/stream.go:101]: Actually, we use write only once and no need for lock here, but libp2p seems so complicated, it's better just to add it and don't worry. Added.

**Comment by Frozen** [comment]
> do you have a testcase that can trigger the concurrency issue?

If node is running with race detector -race, it will fail after some time with that issue.

**Review by rlan35 — APPROVED** [review]

**Pull request merged**